Mathlib/Data/Nat/Factorization/Basic.lean

```diff
@@ -478,4 +478,71 @@ lemma card_multiples' (N n : ℕ) : #{k ∈ range N.succ | k ≠ 0 ∧ n ∣ k}
         · simp [h, succ_div_of_dvd, ih]
         · simp [h, succ_div_of_not_dvd, ih]
 
+theorem exists_eq_pow_of_exponent_coprime_of_pow_eq_pow
+    {a b m n : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) (hmn : m.Coprime n) (h : a ^ m = b ^ n) :
+    ∃ c, a = c ^ n ∧ b = c ^ m := by
+  by_cases hn : n = 0
+  · use b
+    simp_all
+  have factorization_pow_eq := congrArg factorization h
+  rw [factorization_pow, factorization_pow] at factorization_pow_eq
+  let c_factorization := a.factorization.mapRange (· / n) (Nat.zero_div n)
+  let c := c_factorization.prod (· ^ ·)
+  use c
+  have factorization_eq_n_smul_c_factorization_of_eq_c_pow_n
+      x (hx : x ≠ 0) n (h : x.factorization = n • c_factorization) : x = c ^ n := by
+    suffices x.factorization = (c ^ n).factorization by
+      refine eq_of_factorization_eq hx ?_ (congrFun (congrArg DFunLike.coe this))
+      suffices c ≠ 0 from pow_ne_zero n this
+      simp [c, c_factorization]
+    convert h
+    rw [factorization_pow]
+    suffices c.factorization = c_factorization by rw [this]
+    unfold c
+    refine prod_pow_factorization_eq_self ?_
+    intro p p_mem
+    exact prime_of_mem_primeFactors (Finsupp.support_mapRange p_mem)
+  have mul_factorization_p_eq_and_n_dvd_a_factorization_p p :
+      m * a.factorization p = n * b.factorization p ∧ n ∣ a.factorization p := by
+    have := congrFun (congrArg DFunLike.coe factorization_pow_eq) p
```

Comment:
If you want to avoid explicit calls to things like `congrFun`, `congrArg` and friends, using the `congr()` term elaborator is an option:

```suggestion
    have := congr($factorization_pow_eq p)
```